### PR TITLE
Ma/session used at

### PR DESCRIPTION
**Pull request by markusahlstrand**
Added an update function for sessions so that we update the used_at every time a session is renewed.

### CODE REVIEW

File: src/adapters/interfaces/Sessions.ts [hunk 1]
```diff
@@ -3,5 +3,10 @@ import { Session } from "../../types/Session";
 export interface SessionsAdapter {
   create: (session: Session) => Promise<void>;
   get: (tenant_id: string, id: string) => Promise<Session | null>;
+  update: (
+    tenant_id: string,
+    id: string,
+    session: { used_at: string },
```
Comment (danielgent) [inline src/adapters/interfaces/Sessions.ts:9]: :smile: 

With the magic of typescript this will be an `any` with one specific key, `used_at`, right? :smile: 

I think we need a more common pattern here. For each entity we do it differently.  For User we have a `PostUsersBody` that we're not even using for `.create`, _but_ for updating we're doing `Partial<PostUsersBody>`

For some entities we pass in all the fields when calling the create adapter, for others we pass in the minimum and the adapter does the timestamps

I've no opinion really, just that we're consistent :sunglasses:

File: src/adapters/kysely/sessions/create.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ export function create(db: Kysely<Database>) {
       client_id: session.client_id,
       created_at: session.created_at.toISOString(),
       expires_at: session.expires_at.toISOString(),
-      used_at: session.expires_at.toISOString(),
+      used_at: session.used_at.toISOString(),
```
Comment (danielgent) [inline src/adapters/kysely/sessions/create.ts:13]: :fire:

File: src/adapters/kysely/sessions/update.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+export function update(db: Kysely<Database>) {
+  return async (
+    tenant_id: string,
+    id: string,
+    session: { used_at: string },
+  ) => {
+    const results = await db
+      .updateTable("sessions")
+      .set(session)
+      // .where("tenant_id", "=", tenant_id)
```
Comment (danielgent) [inline src/adapters/kysely/sessions/update.ts:13]: but why?  And why leave the comments in? Is this change intentional?

File: src/authentication-flows/silent.ts [hunk 4]
```diff
@@ -77,6 +77,9 @@ export async function silentAuth({
         });
 
         ctx.set("log", JSON.stringify(tokenResponse));
+        await env.data.sessions.update(tenant_id, tokenState, {
+          used_at: new Date().toISOString(),
```
Comment (danielgent) [inline src/authentication-flows/silent.ts:81]: I could see the merit in having a method `setUsedAtNow()`

File: test/integration/flows/silent-auth.spec.ts [hunk 5]
```diff
@@ -47,6 +47,63 @@ describe("silent-auth", () => {
...
+      "nonce",
+      "clientId",
+    );
+    expect(idToken).toBeDefined();
+    const session = await env.data.sessions.get("tenantId", idToken.sid);
+
+    expect(session!.used_at > postLoginDate).toBe(true);
+  });
+
+  it("should set the used_at property on the session when the token is renewed", async () => {
```
Comment (danielgent) [inline test/integration/flows/silent-auth.spec.ts:103]: eh? it looks like you've changed the name of an existing test and named your new test the same... :smile: 

I think we could just bundle this assertion in with the first test.   I know in the 1990s we were told *ONE THING PER TEST* but nowadays the likes of Kent C Dodds say to bundle more in longer flowing tests

That 1990s testing clearly never worked because look at Micke & Nacho :smile:

File: src/authentication-flows/silent.ts (hunk 4 above)
Comment (markusahlstrand) [reply src/authentication-flows/silent.ts:81]: Yes, depending on if we want to be able to set more things in the future.. That naming would be clearer for sure :)

File: src/adapters/kysely/sessions/update.ts (hunk 3 above)
Comment (markusahlstrand) [reply src/adapters/kysely/sessions/update.ts:13]: Debugging stuff left in. Had issues getting this to work and think I tested to comment the tenant out. I'll fix

File: src/adapters/interfaces/Sessions.ts (hunk 1 above)
Comment (markusahlstrand) [reply src/adapters/interfaces/Sessions.ts:9]: Yepp, totally agree. Think we should have this in writing somewhere.

I don't think this will be any any though as it's not using the partial keyword? It would be nice to be able to use Zod.. think that could have solved the partial mess

File: test/integration/flows/silent-auth.spec.ts (hunk 5 above)
Comment (markusahlstrand) [reply test/integration/flows/silent-auth.spec.ts:103]: What is the benefit of just having one big test though? Making it run a bit faster since you don't need to do the setup twice? Maybe we can have one pretty chunky happy-path tests, but thinking that all the negative tests might be better to keep separate?

Comment (markusahlstrand) [reply test/integration/flows/silent-auth.spec.ts:103]: Totally right about changing the name of the wrong test by the way :)

File: src/adapters/interfaces/Sessions.ts (hunk 1 above)
Comment (danielgent) [reply src/adapters/interfaces/Sessions.ts:9]: Yes you're right! :smile: 

I'll re-read that stackoverflow thread you sent me. I thought extra keys was the norm except in a few shorthands :see_no_evil:

**Review by danielgent — APPROVED** [review]

File: test/integration/flows/silent-auth.spec.ts (hunk 5 above)
Comment (danielgent) [reply test/integration/flows/silent-auth.spec.ts]: https://kentcdodds.com/blog/write-fewer-longer-tests wow. almost 5 years old :smiling_face_with_tear:  Where has my life gone?

Just an FYI. Nothing I would say that KCD hasn't already says. I like how he talks about improvements in DX mean we don't need such short tests
